### DISCUSSION TITLE
Made an ebuild

### DISCUSSION THREAD

**Discussion by Schievel1** [Antynea/grub-btrfs#173] [General]
Hello, 

I made an ebuild of grub-btrfs for Gentoo which I am currently testing and would like to add to the GURU (that is the AUR for Gentoo) later. I know its under der GPL3, but I just wanted to let you know and get an ok from you, before I redistribute your software in this way.

**Comment by Antynea** [comment]
Hello,

I am glad to hear that you are interested in this project and that you want to share it with the Gentoo community.
I encourage this initiative.
Let me know when you have finished your work,
i would like to add installation notes on Gentoo in the [readme ](https://github.com/Antynea/grub-btrfs#installation-)file.

**Reply by Schievel1** [reply]
ok I will let you know when they release my ebuild into guru-master. 
Right now I got it installing fine in gentoo with systemd. I have not tested on openRC yet, but I think the basic function will be there. I just need to port that systemd service, that you use to make a new config file for grub everytime there is a new snapshot, into a daemon for openRC.

**Reply by Antynea** [reply]
Hello,

Perhaps you might be interested in [this](https://github.com/oshaughnessyj/grub-btrfs-openrc)

**Reply by Schievel1** [reply]
Thanks for that, I didn’t know somebody is working on this already.
But all he or she did als far as I can see is to make your oneliner with the if function to either update grub.ctg or update grub-btrfs.cfg into a little nicer looking openrc daemon.

As far as I understand your approach you monitor the systemd mount and call the Systemd service with the grub-mkconfig when it is mounted (or when something changes? Idk)
There is no such thing like openrc mounts and therefore the hardest part is to write something that monitors the snapshot directory.


However, For now I just went with a different approach. It is actually very easy to just run a command at system shutdown and system startup. Just put a script with the extension .start or .stop into /etc/local.d, done. 
I will write instructions on that here as soon as the ebuild is released

**Comment by Antynea** [comment]
> As far as I understand your approach you monitor the systemd mount and call the Systemd service with the grub-mkconfig when it is mounted (or when something changes? Idk)

`grub-btrfs.path` monitors the mount point, whenever it is available.
Then if a change occurs inside, it executes `grub-btrfs.service`

When `grub-btrfs.service` is started,
the `PATH` environment variable is temporarily changed for this service. 
(in order to widen the path of the executables of the different distributions)
`/etc/default/grub-btrfs/config` is sourced.
`/boot/grub/grub-btrfs.cfg` is generated either:
With `grub-mkconfig` if the file is empty or non-existent, to insert a custom menu at `/boot/grub.cfg`.
With `/etc/grub.d/41_snapshots-btrfs` if the file exists and is non-empty, to update only the `grub-btrfs.cfg` file.

`{GRUB_BTRFS_GRUB_DIRNAME:-/boot/grub}` corresponds to the customizable variable in the `/etc/default/grub-btrfs/config` file.
`{GRUB_BTRFS_MKCONFIG:-grub-mkconfig}` Same as above.

> There is no such thing like openrc mounts and therefore the hardest part is to write something that monitors the snapshot directory.

`inotify` is not capable of this, monitoring `/proc/mounts` is tedious, `systemd` has been a friend for this feature.
Sorry to hear that there is nothing like it with openrc.

> However, For now I just went with a different approach. It is actually very easy to just run a command at system shutdown and system startup. Just put a script with the extension .start or .stop into /etc/local.d, done.

I thought of this idea some time ago to get around a problem with `grub-btrfs.path`.
I put a note about it in the [readme](https://github.com/Antynea/grub-btrfs/issues/138#issuecomment-766918328).
Make sure your script is robust, because at the slightest error, Goodbye your `Grub menu`.
(It would be a shame if the next time the system boots, Grub is broken)

> I will write instructions on that here as soon as the ebuild is released

Keep up the good work

**Reply by Schievel1** [reply]
Yeah, problem is I do not really understand systemd files yet. I never had anything to do with it, I just have an old laptop with gentoo with systemd on it to test stuff. 

My script in /etc/local.d is basically !#/bin/bash and then this line I stole from your service:
bash -c 'if [ -s "${GRUB_BTRFS_GRUB_DIRNAME:-/boot/grub}/grub-btrfs.cfg" ]; then /etc/grub.d/41_snapshots-btrfs; else ${GRUB_BTRFS_MKCONFIG:-grub-mkconfig} -o ${GRUB_BTRFS_GRUB_DIRNAME:-/boot/grub}/grub.cfg; fi' 

so as long as your script works, I think I'm fine :D

**Reply by Schievel1** [reply]
> `inotify` is not capable of this, monitoring `/proc/mounts` is tedious, `systemd` has been a friend for this feature. Sorry to hear that there is nothing like it with openrc.

This may be naive, but why is inotify not capable?

`#!/sbin/openrc-run

start() {
     . /etc/default/grub-btrfs/config
     while true; do sleep 1 && inotifywait -e create /run/timeshift/backup/timeshift-btrfs/snapshots && sleep 5 && if [ -s "${GRUB_BTRFS_GRUB_DIRNAME:-/boot/grub}/grub-btrfs.cfg" ]; then /etc/grub.d/41_snapshots-btrfs; else ${GRUB_BTRFS_MKCONFIG:-grub-mkconfig} -o ${GRUB_BTRFS_GRUB_DIRNAME:-/boot/grub}/grub.cfg; fi  ; done
}`

This is what I got so far. It works, but I do not really like running inotifywait every second until timeshift is opened once and the directory becomes available.

/ just in case anyone reads this an tries this script: do not add this to your default runlevel! this will probably make your system unbootable!

**Reply by Antynea** [reply]
> There is no such thing like openrc mounts and therefore the hardest part is to write something that monitors the snapshot directory.
>>inotify is not capable of this, monitoring /proc/mounts is tedious, systemd has been a friend for this feature. Sorry to hear that there is nothing like it with openrc.
>>> This may be naive, but why is inotify not capable?

If you monitor an existing folder with `inotify` before the subvolume or device is mounted, `inotify` will continue to monitor an empty folder.

**Reply by Schievel1** [reply]
> > There is no such thing like openrc mounts and therefore the hardest part is to write something that monitors the snapshot directory.
> > > inotify is not capable of this, monitoring /proc/mounts is tedious, systemd has been a friend for this feature. Sorry to hear that there is nothing like it with openrc.
> > > > This may be naive, but why is inotify not capable?
> 
> If you monitor an existing folder with `inotify` before the subvolume or device is mounted, `inotify` will continue to monitor an empty folder.

But the folder doesnt exist after boot. Only timeshift runs for the first time, the folder /run/timeshift/backup/snapshots becomes available. 
So inotifywait just gives me an error massage. I can live with that an try again in 5 seconds.

I got it running as a working openRC service with my approach from above. Nevertheless, I am not happy with the polling. Some kind of a callback would be a cleaner solution.

**Comment by Kr1ss-XD** [comment]
Hi @Schievel1 - thanks a lot for this effort.

I'd like to mention that it is possible to initiate commands on filesystem changes, without having to rely on systemd facilities. Maybe [fswatch](https://github.com/emcrisostomo/fswatch) would be worth a look, although I'm not sure it this wouldn't be overkill in this case.  In any case, on Linux, it can watch the snapshot folder either by inotify or by polling, and both seems to work fine here.

There are some similar projects out there, maybe you can find something more minimal and/or Linux specific, which you could run as a daemon.

_Cheers !_

**Comment by Schievel1** [comment]
Ebuild is in guru finally.
Feel free to add this to your README.md or where ever you like:
Before you ask: I am not allowed to do this initial grub-mkconfig automatically during install. That is why I would like to add it into the instructions here. I also call this out to the user as a log massage at the end of of the install, but portage is quite verbose to say the least, so its easy to miss.  

**Instructions for Gentoo:**
grub-btrfs is only available in the Gentoo User Repository (GURU) and not in the official Gentoo repository.
If you have not activated the GURU yet, do so by running
```
emerge -av app-eselect/eselect-repository 
eselect repository enable guru 
emerge --sync 
```

Now merge grub-btrfs via 
`emerge app-backup/grub-btrfs`

After installation run 
`grub-mkconfig -o /boot/grub/grub.cfg `
to initially add the snapshots you already have to your grub menu. 

Once it added its own submenu to grub, its enough to run /etc/grub.d/41_snapshots-btrfs
to add new snapshots to the grub submenu. 

**systemd**

If you are using gentoo with systemd, you can use the systemd-service provided by the grub-btrfs package to monitor /run/timeshift/backup/timeshift-btrfs/snapshots. 
***probably link to your systemd service instructions here***

**openRC**

If your are on openRC instead, there is sadly no similar solution to the systemd service yet. 
As a workaround it is possible to add a script to /etc/local.d to execute the snapshot menu update every time you shut down your system.
To do so just add the following script as /etc/local.d/grub-btrfs-update.stop
```
#!/bin/bash

description="Update the grub btrfs snapshots menu"
name="grub-btrfs-update"

depend()
{
  use localmount
}

bash -c 'if [ -s "${GRUB_BTRFS_GRUB_DIRNAME:-/boot/grub}/grub-btrfs.cfg" ]; then /etc/grub.d/41_snapshots-btrfs; else {GRUB_BTRFS_MKCONFIG:-grub-mkconfig} -o {GRUB_BTRFS_GRUB_DIRNAME:-/boot/grub}/grub.cfg; fi' 
```

If you want to run the menu update on startup instead, rename the file to grub-btrfs-update.start

**Reply by Schievel1** [reply]
Could you add:

"Make your script executeable with `chmod a+x /etc/local.d/grub-btrfs-update.stop`."

to the end of this? I dont want to make a full fork and pull request and everything just for that

**Reply by Antynea** [reply]
I can add this on `Master branch` now.
I realise that you did this in your PR.
Would you still like me to do it now ?
This will require your #178 to be up to date with the master branch again before merging, which is not ideal.

**Reply by Schievel1** [reply]
if you merge the PR its fine

**Reply by Antynea** [reply]
Done.
